core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -27,15 +57,44 @@ public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
       }
       try {
          final int commandTopologyId = extractCommandTopologyId(command);
-         final boolean waitTransactionalData = command.getCommandId() != StateRequestCommand.COMMAND_ID;
          final boolean onExecutorService = executeOnExecutorService(order, command);
-         handleRunnable(createDefaultRunnable(command, reply, commandTopologyId, waitTransactionalData, onExecutorService), onExecutorService);
+         final BlockingRunnable runnable;
+
+         switch (command.getCommandId()) {
+            case SingleRpcCommand.COMMAND_ID:
+               //lock ownership needs the first topology. we need to check it before invoking the getLockPromise()
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
```

Comment:
I can do something similar to https://github.com/infinispan/infinispan/pull/3606/files#diff-376c88e74f4d66c0dff57dc91b234d89R356

when isReady() is invoked, I check if topology is received, then the first topology and in the end, acquire the check LockPromise. do you have another idea?